### ISSUE TITLE
Kind unification bug

### ISSUE THREAD

**Issue by kcsongor** [purescript/purescript#3077]
As discussed on slack, there seems to be an issue with how kinds are unified, here's a minimal, self-contained example:

```purescript
data TProxy (t :: Type) = TProxy
data SProxy (s :: Symbol) = SProxy

put :: forall proxy a. proxy a -> TProxy a
put _ = TProxy

--wrong :: TProxy "apple"
wrong = put (SProxy :: SProxy "apple")
```

This typechecks (on both HEAD and 0.11.6), and we managed to force a `Symbol` into `TProxy`.

**Comment by paf31** [comment]
I think the problem here is that only the kinds of written types are checked. The only incorrect type is the type of `wrong`, which isn't checked.

**Comment by paf31** [comment]
I made a quick attempt to fix this, but it's a bit trickier than I thought. I think it actually needs some of the same changes that we would need for polykinds, namely adding kind information to the types of values.

**Comment by rhendric** [comment]
Looks like #3779 fixed this; I could not reproduce on v0.14.
